LDAP authentication fails with ldaps server

I want to use my university's LDAP server (anonymized below as authorise.uni.ac.uk) for OpenCGA authentication. The LDAP server is accessed via ldaps protocol on port 636. It appears that the URL is not being correctly parsed - it comes out as '/authorise.uni.ac.uk' instead of 'authorise.uni.ac.uk' in the error messages. When I use ldap:// instead of ldaps:// the host name is correctly identified, but won't connect as it's the wrong port/protocol combination.

configuration.yml:
```
authentication:
  # Session expiration time in seconds
  expiration: 3600
# LDAP configuration example
  authenticationOrigins:
  - id: UNI_LDAP            # Any id
    type: LDAP
    host: ldaps://authorise.uni.ac.uk:636
    options:
      usersSearch: dc=authorise,dc=uni,dc=ac,dc=uk # Base search to look for the users
      groupsSearch: ou=people,ou=central,dc=authorise,dc=uni,dc=ac,dc=uk # Base search to look for the groups
```
Error:
```
[opencga@server conf]$ opencga-admin.sh users import --auth-origin UNI_LDAP -u ameyner2
Admin password: 
2019-04-15 20:31:01 [main] ERROR LDAPAuthenticationManager:93 - Invalid name: /authorise.uni.ac.uk:636
javax.naming.InvalidNameException: Invalid name: /authorise.uni.ac.uk:636
	at javax.naming.ldap.Rfc2253Parser.parseAttrType(Rfc2253Parser.java:155)
	at javax.naming.ldap.Rfc2253Parser.doParse(Rfc2253Parser.java:108)
	at javax.naming.ldap.Rfc2253Parser.parseDn(Rfc2253Parser.java:70)
	at javax.naming.ldap.LdapName.parse(LdapName.java:785)
	at javax.naming.ldap.LdapName.<init>(LdapName.java:123)
	at com.sun.jndi.ldap.LdapNameParser.parse(LdapNameParser.java:39)
	at com.sun.jndi.ldap.LdapCtx.<init>(LdapCtx.java:302)
	at com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:192)
	at com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(LdapCtxFactory.java:210)
	at com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxInstance(LdapCtxFactory.java:153)
	at com.sun.jndi.ldap.LdapCtxFactory.getInitialContext(LdapCtxFactory.java:83)
	at javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:684)
	at javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:313)
	at javax.naming.InitialContext.init(InitialContext.java:244)
	at javax.naming.InitialContext.<init>(InitialContext.java:216)
	at javax.naming.directory.InitialDirContext.<init>(InitialDirContext.java:101)
	at org.opencb.opencga.catalog.auth.authentication.LDAPUtils.getDirContext(LDAPUtils.java:48)
	at org.opencb.opencga.catalog.auth.authentication.LDAPUtils.getUserInfoFromLDAP(LDAPUtils.java:144)
	at org.opencb.opencga.catalog.auth.authentication.LDAPAuthenticationManager.authenticate(LDAPAuthenticationManager.java:74)
	at org.opencb.opencga.catalog.managers.UserManager.login(UserManager.java:506)
	at org.opencb.opencga.app.cli.admin.executors.UsersCommandExecutor.importUsersAndGroups(UsersCommandExecutor.java:133)
	at org.opencb.opencga.app.cli.admin.executors.UsersCommandExecutor.execute(UsersCommandExecutor.java:61)
	at org.opencb.opencga.app.cli.admin.AdminMain.main(AdminMain.java:129)
2019-04-15 20:31:01 [main] INFO  UserManager:331 - Fetching user information from authentication origin 'UNI_LDAP'
2019-04-15 20:31:04 [main] ERROR LDAPAuthenticationManager:121 - Could not retrieve user information Invalid name: /authorise.uni.ac.uk:636
{}
javax.naming.InvalidNameException: Invalid name: /authorise.uni.ac.uk:636
	at javax.naming.ldap.Rfc2253Parser.parseAttrType(Rfc2253Parser.java:155)
	at javax.naming.ldap.Rfc2253Parser.doParse(Rfc2253Parser.java:108)
	at javax.naming.ldap.Rfc2253Parser.parseDn(Rfc2253Parser.java:70)
	at javax.naming.ldap.LdapName.parse(LdapName.java:785)
	at javax.naming.ldap.LdapName.<init>(LdapName.java:123)
	at com.sun.jndi.ldap.LdapNameParser.parse(LdapNameParser.java:39)
	at com.sun.jndi.ldap.LdapCtx.<init>(LdapCtx.java:302)
	at com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:192)
	at com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(LdapCtxFactory.java:210)
	at com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxInstance(LdapCtxFactory.java:153)
	at com.sun.jndi.ldap.LdapCtxFactory.getInitialContext(LdapCtxFactory.java:83)
	at javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:684)
	at javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:313)
	at javax.naming.InitialContext.init(InitialContext.java:244)
	at javax.naming.InitialContext.<init>(InitialContext.java:216)
	at javax.naming.directory.InitialDirContext.<init>(InitialDirContext.java:101)
	at org.opencb.opencga.catalog.auth.authentication.LDAPUtils.getDirContext(LDAPUtils.java:48)
	at org.opencb.opencga.catalog.auth.authentication.LDAPUtils.getUserInfoFromLDAP(LDAPUtils.java:144)
	at org.opencb.opencga.catalog.auth.authentication.LDAPAuthenticationManager.getRemoteUserInformation(LDAPAuthenticationManager.java:119)
	at org.opencb.opencga.catalog.managers.UserManager.importRemoteUsers(UserManager.java:333)
	at org.opencb.opencga.app.cli.admin.executors.UsersCommandExecutor.importUsersAndGroups(UsersCommandExecutor.java:136)
	at org.opencb.opencga.app.cli.admin.executors.UsersCommandExecutor.execute(UsersCommandExecutor.java:61)
	at org.opencb.opencga.app.cli.admin.AdminMain.main(AdminMain.java:129)
org.opencb.opencga.catalog.exceptions.CatalogException: Could not retrieve user information
Invalid name: /authorise.uni.ac.uk:636
	at org.opencb.opencga.catalog.auth.authentication.LDAPAuthenticationManager.getRemoteUserInformation(LDAPAuthenticationManager.java:122)
	at org.opencb.opencga.catalog.managers.UserManager.importRemoteUsers(UserManager.java:333)
	at org.opencb.opencga.app.cli.admin.executors.UsersCommandExecutor.importUsersAndGroups(UsersCommandExecutor.java:136)
	at org.opencb.opencga.app.cli.admin.executors.UsersCommandExecutor.execute(UsersCommandExecutor.java:61)
	at org.opencb.opencga.app.cli.admin.AdminMain.main(AdminMain.java:129)
Caused by: javax.naming.InvalidNameException: Invalid name: /authorise.uni.ac.uk:636
	at javax.naming.ldap.Rfc2253Parser.parseAttrType(Rfc2253Parser.java:155)
	at javax.naming.ldap.Rfc2253Parser.doParse(Rfc2253Parser.java:108)
	at javax.naming.ldap.Rfc2253Parser.parseDn(Rfc2253Parser.java:70)
	at javax.naming.ldap.LdapName.parse(LdapName.java:785)
	at javax.naming.ldap.LdapName.<init>(LdapName.java:123)
	at com.sun.jndi.ldap.LdapNameParser.parse(LdapNameParser.java:39)
	at com.sun.jndi.ldap.LdapCtx.<init>(LdapCtx.java:302)
	at com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:192)
	at com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(LdapCtxFactory.java:210)
	at com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxInstance(LdapCtxFactory.java:153)
	at com.sun.jndi.ldap.LdapCtxFactory.getInitialContext(LdapCtxFactory.java:83)
	at javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:684)
	at javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:313)
	at javax.naming.InitialContext.init(InitialContext.java:244)
	at javax.naming.InitialContext.<init>(InitialContext.java:216)
	at javax.naming.directory.InitialDirContext.<init>(InitialDirContext.java:101)
	at org.opencb.opencga.catalog.auth.authentication.LDAPUtils.getDirContext(LDAPUtils.java:48)
	at org.opencb.opencga.catalog.auth.authentication.LDAPUtils.getUserInfoFromLDAP(LDAPUtils.java:144)
	at org.opencb.opencga.catalog.auth.authentication.LDAPAuthenticationManager.getRemoteUserInformation(LDAPAuthenticationManager.java:119)
	... 4 more
```

--- a/opencga-catalog/src/main/java/org/opencb/opencga/catalog/auth/authentication/LDAPAuthenticationManager.java
+++ b/opencga-catalog/src/main/java/org/opencb/opencga/catalog/auth/authentication/LDAPAuthenticationManager.java
@@ -42,21 +42,34 @@ public class LDAPAuthenticationManager extends AuthenticationManager {
 
     private String groupsSearch;
     private String usersSearch;
+    private String fullNameKey;
+    private String dnKey;
+    private String dnFormat;
 
     private long expiration;
 
     private static final String OPENCGA_GECOS = "opencga_gecos";
+    private Boolean ldaps;
 
     public LDAPAuthenticationManager(AuthenticationOrigin authenticationOrigin, String secretKeyString, long expiration) {
         super();
 
         this.host = authenticationOrigin.getHost();
-        if (!this.host.startsWith("ldap://")) {
+
+        if (host.startsWith("ldaps://")) {  // use LDAPS if specified explicitly
+            this.ldaps = true;
+        } else if (!this.host.startsWith("ldap://")) {  // otherwise default to LDAP
             this.host = "ldap://" + this.host;
         }
+
         this.originId = authenticationOrigin.getId();
-        this.groupsSearch = String.valueOf(authenticationOrigin.getOptions().get(AuthenticationOrigin.GROUPS_SEARCH));
-        this.usersSearch = String.valueOf(authenticationOrigin.getOptions().get(AuthenticationOrigin.USERS_SEARCH));
+
+        Map<String, Object> authOptions = authenticationOrigin.getOptions();
+        this.groupsSearch = String.valueOf(authOptions.get(AuthenticationOrigin.GROUPS_SEARCH));
+        this.usersSearch = String.valueOf(authOptions.get(AuthenticationOrigin.USERS_SEARCH));
+        this.fullNameKey = String.valueOf(authOptions.get(AuthenticationOrigin.FULLNAME_KEY));
+        this.dnKey = String.valueOf(authOptions.get(AuthenticationOrigin.DN_KEY));
+        this.dnFormat = String.valueOf(authOptions.get(AuthenticationOrigin.DN_FORMAT));
 
         this.expiration = expiration;
 
@@ -84,8 +97,16 @@ public class LDAPAuthenticationManager extends AuthenticationManager {
             env.put(Context.PROVIDER_URL, host);
 
             env.put(Context.SECURITY_AUTHENTICATION, "simple");
-            env.put(Context.SECURITY_PRINCIPAL, rdn);
             env.put(Context.SECURITY_CREDENTIALS, password);
+
+            if (ldaps) {
+                String uid = userInfoFromLDAP.get(0).get(dnKey).get(0).toString();
+                String dn = String.format(dnFormat, uid);
+                env.put(Context.SECURITY_PRINCIPAL, dn);
+                env.put(Context.SECURITY_PROTOCOL, "ssl");
+            } else {
+                env.put(Context.SECURITY_PRINCIPAL, rdn);
+            }
 
             // Create the initial context
             new InitialDirContext(env);
@@ -134,7 +155,7 @@ public class LDAPAuthenticationManager extends AuthenticationManager {
         String rdn;
         for (Attributes attrs : userAttrList) {
             try {
-                displayName = LDAPUtils.getFullName(attrs);
+                displayName = LDAPUtils.getFullName(attrs, fullNameKey);
                 mail = LDAPUtils.getMail(attrs);
                 uid = LDAPUtils.getUID(attrs);
                 rdn = LDAPUtils.getRDN(attrs);

--- a/opencga-catalog/src/main/java/org/opencb/opencga/catalog/auth/authentication/LDAPAuthenticationManager.java
+++ b/opencga-catalog/src/main/java/org/opencb/opencga/catalog/auth/authentication/LDAPAuthenticationManager.java
@@ -67,9 +67,9 @@ public class LDAPAuthenticationManager extends AuthenticationManager {
         Map<String, Object> authOptions = authenticationOrigin.getOptions();
         this.groupsSearch = String.valueOf(authOptions.get(AuthenticationOrigin.GROUPS_SEARCH));
         this.usersSearch = String.valueOf(authOptions.get(AuthenticationOrigin.USERS_SEARCH));
-        this.fullNameKey = String.valueOf(authOptions.get(AuthenticationOrigin.FULLNAME_KEY));
-        this.dnKey = String.valueOf(authOptions.get(AuthenticationOrigin.DN_KEY));
-        this.dnFormat = String.valueOf(authOptions.get(AuthenticationOrigin.DN_FORMAT));
+        this.fullNameKey = String.valueOf(authOptions.getOrDefault(AuthenticationOrigin.FULLNAME_KEY, "displayname"));
+        this.dnKey = String.valueOf(authOptions.getOrDefault(AuthenticationOrigin.DN_KEY, "gecos"));
+        this.dnFormat = String.valueOf(authOptions.getOrDefault(AuthenticationOrigin.DN_FORMAT, "%s"));  // no formatting by default
 
         this.expiration = expiration;
 

--- a/opencga-catalog/src/main/java/org/opencb/opencga/catalog/auth/authentication/LDAPUtils.java
+++ b/opencga-catalog/src/main/java/org/opencb/opencga/catalog/auth/authentication/LDAPUtils.java
@@ -167,8 +167,8 @@ public class LDAPUtils {
         }
     }
 
-    public static String getFullName(Attributes attributes) throws NamingException {
-        return (String) attributes.get("displayname").get(0);
+    public static String getFullName(Attributes attributes, String fullNameKey) throws NamingException {
+        return (String) attributes.get(fullNameKey).get(0);
     }
 
     public static List<String> getGroupsFromLdapUser(String host, String user, String base) throws NamingException {

--- a/opencga-core/src/main/java/org/opencb/opencga/core/config/AuthenticationOrigin.java
+++ b/opencga-core/src/main/java/org/opencb/opencga/core/config/AuthenticationOrigin.java
@@ -38,9 +38,9 @@ public class AuthenticationOrigin {
     // Possible keys of the options map
     public static final String USERS_SEARCH = "usersSearch";
     public static final String GROUPS_SEARCH = "groupsSearch";
-    public static final String FULLNAME_KEY = "fullNameKey";  // key to get user's full name, e.g. "displayname"
-    public static final String DN_KEY = "dnKey";  // key for user's unique ID, e.g. "uid", "gecos"
-    public static final String DN_FORMAT = "dnFormat";  // distinguished name format string
+    public static final String FULLNAME_KEY = "fullNameKey";
+    public static final String DN_KEY = "dnKey";
+    public static final String DN_FORMAT = "dnFormat";
 
     public AuthenticationOrigin() {
         this("internal", AuthenticationType.OPENCGA.name(), "localhost", Collections.emptyMap());

--- a/opencga-core/src/main/java/org/opencb/opencga/core/config/AuthenticationOrigin.java
+++ b/opencga-core/src/main/java/org/opencb/opencga/core/config/AuthenticationOrigin.java
@@ -38,6 +38,9 @@ public class AuthenticationOrigin {
     // Possible keys of the options map
     public static final String USERS_SEARCH = "usersSearch";
     public static final String GROUPS_SEARCH = "groupsSearch";
+    public static final String FULLNAME_KEY = "fullNameKey";  // key to get user's full name, e.g. "displayname"
+    public static final String DN_KEY = "dnKey";  // key for user's unique ID, e.g. "uid", "gecos"
+    public static final String DN_FORMAT = "dnFormat";  // distinguished name format string
 
     public AuthenticationOrigin() {
         this("internal", AuthenticationType.OPENCGA.name(), "localhost", Collections.emptyMap());

--- a/opencga-core/src/main/resources/configuration.yml
+++ b/opencga-core/src/main/resources/configuration.yml
@@ -47,6 +47,18 @@ authentication:
 #    options:
 #      usersSearch: dc=ge,dc=co,dc=uk # Base search to look for the users
 #      groupsSearch: ou=general,ou=groups,dc=ge,dc=co,dc=uk # Base search to look for the groups
+
+# Custom LDAPS configuration example
+#  - id: ldaps            # Any id
+#    type: LDAP
+#    host: ldaps://localhost:636
+#    options:
+#      usersSearch: dc=ge,dc=co,dc=uk                       # Base search to look for users
+#      groupsSearch: ou=general,ou=groups,dc=ge,dc=co,dc=uk # Base search to look for groups
+#      fullNameKey: displayname                             # Key to get the user's full name when importing users
+#      dnKey: gecos                                         # Get the user's unique identifier...
+#      dnFormat: "uid=%s,ou=people,dc=ge,dc=co,dc=uk"       # ...and pass a distinguished name to the auth request in the right format
+
 # Azure AD configuration example
 #  - id: aad                                               # Any id
 #    type: AzureAD

